### ISSUE TITLE
rnpkeys with a large argument in --homedir causes a crash

### ISSUE THREAD

**Issue by erikbor** [riboseinc/rnp#437]
## Description

can crash rnpkeys with a long argument in --homedir

## Steps to Reproduce

```sh
$ uname -a
Darwin OF-OSX-007 16.6.0 Darwin Kernel Version 16.6.0: Fri Apr 14 16:21:16 PDT 2017; root:xnu-3789.60.24~6/RELEASE_X86_64 x86_64
$ rnpkeys --list-keys --homedir=$(perl -e 'print "A"x1024')
Abort trap: 6 (core dumped)
```

## Expected Behavior

No crash

## Actual Behavior

A core dump

**Comment by ronaldtse** [comment]
Ah, @erikbor 's favorite behavior with a working solution from @jjr840430 😉

**Comment by jjr840430** [comment]
@erikbor My account is denied to push the changes.

Here is the patch to fix this issue.
```
diff --git a/src/rnp/rnpcfg.c b/src/rnp/rnpcfg.c
index 9155529..f92e056 100644
--- a/src/rnp/rnpcfg.c
+++ b/src/rnp/rnpcfg.c
@@ -411,7 +411,7 @@ conffile(const char *homedir, char *userid, size_t length)
  *  @param dir [in] null-terminated directory path, cannot be NULL
  *  @param subddir [in] null-terminated subdirectory to add to the path, can be NULL
  *  @param filename [in] null-terminated filename (or path/filename), cannot be NULL
- *  @param res [out] preallocated buffer, large enough to store the result
+ *  @param res [out] preallocated buffer
  *
  *  @return true if path constructed successfully, or false otherwise
  **/
@@ -426,6 +426,9 @@ rnp_path_compose(const char *dir, const char *subdir, const char *filename, char
     }
 
     /* concatenating dir, subdir and filename */
+    if (strlen(dir) > MAXPATHLEN - 1)
+        return false;
+
     strcpy(res, dir);
     pos = strlen(dir);
 
@@ -434,6 +437,9 @@ rnp_path_compose(const char *dir, const char *subdir, const char *filename, char
             res[pos++] = '/';
         }
 
+        if (strlen(subdir) + pos > MAXPATHLEN - 1)
+            return false;
+
         strcpy(res + pos, subdir);
         pos += strlen(subdir);
     }
@@ -442,6 +448,9 @@ rnp_path_compose(const char *dir, const char *subdir, const char *filename, char
         res[pos++] = '/';
     }

```

**Comment by ronaldtse** [comment]
@jjr840430 Permission granted. Next time you could also do a PR from your own fork, which doesn't require push permissions.

**Comment by jjr840430** [comment]
Fixed by #444
